crates/ruma-common/src/events/image.rs

```diff
@@ -0,0 +1,333 @@
+    /// Creates a new `Captions` with the given captions.
+    ///
+    /// The captions must be ordered by most preferred first.
+    pub fn new(captions: &[Text]) -> Self {
+        Self(captions.to_owned())
+    }
+
+    /// A convenience constructor to create a plain text caption.
+    pub fn plain(body: impl Into<String>) -> Self {
+        Self(vec![Text::plain(body)])
+    }
+
+    /// A convenience constructor to create an HTML caption.
+    pub fn html(body: impl Into<String>, html_body: impl Into<String>) -> Self {
+        Self(vec![Text::html(html_body), Text::plain(body)])
+    }
+
+    /// A convenience constructor to create a Markdown caption.
+    ///
+    /// Returns an HTML caption if some Markdown formatting was detected, otherwise returns a plain
+    /// text caption.
+    #[cfg(feature = "markdown")]
+    pub fn markdown(body: impl AsRef<str> + Into<String>) -> Self {
+        let mut message = Vec::with_capacity(2);
+        if let Some(html_body) = Text::markdown(&body) {
+            message.push(html_body);
+        }
+        message.push(Text::plain(body));
+        Self(message)
+    }
+
+    /// Get the captions.
+    ///
+    /// The captions are ordered by most preferred first.
+    pub fn captions(&self) -> &[Text] {
+        &self.0
+    }
+
+    /// Whether this is empty.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Get the plain text representation of this caption.
+    pub fn find_plain(&self) -> Option<&str> {
+        self.captions()
+            .iter()
+            .find(|content| content.mimetype == "text/plain")
+            .map(|content| content.body.as_ref())
+    }
+
+    /// Get the HTML representation of this caption.
+    pub fn find_html(&self) -> Option<&str> {
+        self.captions()
+            .iter()
+            .find(|content| content.mimetype == "text/html")
+            .map(|content| content.body.as_ref())
+    }
+}
+
+/// Image content.
+#[derive(Default, Clone, Debug, Serialize, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct ImageContent {
+    /// The height of the image in pixels.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub height: Option<UInt>,
+
+    /// The width of the image in pixels.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub width: Option<UInt>,
+}
+
+impl ImageContent {
+    /// Creates a new empty `ImageContent`.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Creates a new `ImageContent` with the given width and height.
+    pub fn with_size(width: UInt, height: UInt) -> Self {
+        Self { height: Some(height), width: Some(width) }
+    }
+}
+
+/// The payload for an extensible image message.
+#[derive(Clone, Debug, Serialize, Deserialize, EventContent)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[ruma_event(type = "m.image", kind = MessageLike)]
+pub struct ImageEventContent {
```

Comment:
I think it would be nice if the actual event content type was always the first thing in the event module.